substrate/validator-sets/pallet/src/lib.rs

```diff
@@ -1025,6 +1145,95 @@ pub mod pallet {
     }
   }
 
+  impl<T: Config, D: AsRef<[u8]>> KeyOwnerProofSystem<(KeyTypeId, D)> for Pallet<T> {
+    type Proof = MembershipProof;
+    type IdentificationTuple = Public;
+
+    fn prove(key: (KeyTypeId, D)) -> Option<Self::Proof> {
+      let (_, data) = key;
+      let validator = Public::try_from(data.as_ref()).ok()?;
+
+      // TODO: do we want this check here? this is node that wants to send the proof anyway.
+      if !Self::can_slash_serai_validator(&validator) {
+        return None;
+      }
+
+      Some(MembershipProof {
+        session: Self::session(NetworkId::Serai).unwrap().0,
+        // TODO: since this isn't even used, should we just 0 or 1ed out?
+        validator_count: Participants::<T>::get(NetworkId::Serai)
+          .expect("no Serai participants on key prove.")
+          .len() as u32,
+        key: validator,
+      })
+    }
+
+    fn check_proof(key: (KeyTypeId, D), proof: Self::Proof) -> Option<Self::IdentificationTuple> {
+      let (_, data) = key;
+      let validator = Public::try_from(data.as_ref()).ok()?;
+
+      // check the offender and the proof offender are the same.
+      if validator != proof.key {
+        return None;
+      }
+
+      // check validator is valid
+      if !Self::can_slash_serai_validator(&validator) {
+        return None;
+      }
+
+      Some(validator)
+    }
+  }
+
+  impl<T: Config> ReportOffence<Public, Public, BabeEquivocationOffence<Public>> for Pallet<T> {
+    fn report_offence(
+      _: Vec<Public>,
+      offence: BabeEquivocationOffence<Public>,
+    ) -> Result<(), OffenceError> {
+      // slash the offender
+      let offender = offence.offender;
+      Self::slash_serai_validator(&offender);
+
+      // save the offence
+      let key = (offender, u64::from(offence.slot));
+      BabeOffences::<T>::set(key, Some(()));
+      Ok(())
+    }
+
+    fn is_known_offence(
+      offenders: &[Public],
+      slot: &<BabeEquivocationOffence<Public> as Offence<Public>>::TimeSlot,
+    ) -> bool {
+      let key = (offenders[0], u64::from(*slot));
+      Self::babe_offences(key).is_some()
+    }
+  }
+
+  impl<T: Config> ReportOffence<Public, Public, GrandpaEquivocationOffence<Public>> for Pallet<T> {
+    fn report_offence(
+      _: Vec<Public>,
+      offence: GrandpaEquivocationOffence<Public>,
+    ) -> Result<(), OffenceError> {
+      // slash the offender
+      let offender = offence.offender;
+      Self::slash_serai_validator(&offender);
+
+      // save the offence
+      let key = (offender, offence.time_slot.set_id, offence.time_slot.round);
+      GrandpaOffences::<T>::set(key, Some(()));
+      Ok(())
+    }
+
+    fn is_known_offence(
+      offenders: &[Public],
+      slot: &<GrandpaEquivocationOffence<Public> as Offence<Public>>::TimeSlot,
+    ) -> bool {
+      let key = (offenders[0], slot.set_id, slot.round);
+      Self::grandpa_offences(key).is_some()
```

Comment:
This is related to https://github.com/serai-dex/serai/pull/468#discussion_r1418974452. If the offender still in the set and can make blocks or decide on the blocks(in grandpa) then we can have the same guy be the offender twice although since we slash everything the first time, we can't do anything(so why care?) the second time.